cmake/cpu_arch_flags.cmake

```diff
@@ -0,0 +1,62 @@
+#[[
+function(target_compiler_flags target)
+  cmake_parse_arguments(
+    TARGET_COMPILER_FLAGS
+    ""
+    "COMPILER"
+    "PUBLIC;PRIVATE"
+    ${ARGN}
+    )
+  if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL ${TARGET_COMPILER_FLAGS_COMPILER})
+    target_compile_options(${target}
+      PRIVATE ${TARGET_COMPILER_FLAGS_PRIVATE}
+      PUBLIC ${TARGET_COMPILER_FLAGS_PUBLIC})
+  endif()
+endfunction()
+]]
+
+
+function(get_arch_specific_flags architecture compiler)
+
+    # Westmere cpu architecture
+    if ("${ARCH}" STREQUAL "wsm")
+        set(CPU_ARCH_SPECIFIC_FLAGS "-msse3" PARENT_SCOPE)
+    
+    # Sandy Bridge cpu architecture
+    elseif ("${ARCH}" STREQUAL "snb")
+        set(CPU_ARCH_SPECIFIC_FLAGS "-mavx" PARENT_SCOPE)
+    
+    # Haswell cpu architecture
+    elseif ("${ARCH}" STREQUAL "hsw")
+ 
+        if (compiler STREQUAL "Intel")
+            set(CPU_ARCH_SPECIFIC_FLAGS "-xCORE-AVX2" "-fma" PARENT_SCOPE)
+        elseif(compiler STREQUAL "GNU")
+            set(CPU_ARCH_SPECIFIC_FLAGS "-mavx2" "-mfma" PARENT_SCOPE)
+        endif()
+
+    # Knights Corner (Xeon Phi)
+    elseif ("${ARCH}" STREQUAL "knc")
+        set(CPU_ARCH_SPECIFIC_FLAGS "-mmic" "-fma" PARENT_SCOPE)
+    
+    # Knight Landing (Xeon Phi)
+    elseif ("${ARCH}" STREQUAL "knl")
+ 
+        if (compiler STREQUAL "Intel")
+            set(CPU_ARCH_SPECIFIC_FLAGS "-xMIC-AVX512" "-fma" PARENT_SCOPE)
+        elseif(compiler STREQUAL "GNU")
+            set(CPU_ARCH_SPECIFIC_FLAGS "-mavx512f" "-mavx512cd" "-mavx512pf" "-mavx512er" "-mfma" PARENT_SCOPE)
+        endif()
+    
+    # Skylake cpu architecture
+    elseif ("${ARCH}" STREQUAL "skx")
+
+        if (compiler STREQUAL "Intel")
+            set(CPU_ARCH_SPECIFIC_FLAGS "-xMIC-AVX512" "-fma" PARENT_SCOPE)
+        elseif(compiler STREQUAL "GNU")
+            set(CPU_ARCH_SPECIFIC_FLAGS "-march=skylake-avx512" PARENT_SCOPE)
+        endif()
+    
+    endif()
+
+endfunction()
```

Comment:
Can you please remove some of the white space here?
I'm not that happy with all these if statements - it's a bit hard to read. Especially, when we might add more compilers at some point.
Maybe this could be rewritten as a generator expression. Not sure if that's cleaner though...
Another option could be switching the if statements, i.e. first check compilers and then check the architecture?

Reply:
Also maybe we can rename the function to get_arch_flags?